packages/foundry/contracts/test/other/tether/TetherJoin.t.sol

```diff
@@ -0,0 +1,285 @@
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.transfer(address(join), unit);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+
+    function testJoinCombine() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.approve(address(join), unit / 2);
+        vm.prank(user);
+        tether.transfer(address(join), unit / 2);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+}
+
+abstract contract WithFees is Deployed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // enable fees
+        vm.prank(tetherMultiSig);
+        tether.setParams(19, 49);    // maximum
+    }
+}
+
+contract WithFeesTest is WithFees {
+    function testJoinPullWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units * unit / (unit - fee); // scale up first so precision isn't lost
+
+        vm.prank(user);
+        tether.approve(address(join), feeAdjustedUnits);
+        vm.prank(ladle);
+        join.join(user, uint128(feeAdjustedUnits));
+
+        assertTrackMinusEq("userBalance", feeAdjustedUnits, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", units, join.storedBalance());
+        assertTrackPlusEq("joinBalance", units, tether.balanceOf(address(join)));
+    }
+
+    function testJoinPushWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units * unit / (unit - fee); // scale up first so precision isn't lost
+
+        vm.prank(user);
+        tether.transfer(address(join), feeAdjustedUnits);
+        vm.prank(ladle);
+        join.join(user, uint128(units));
+
+        assertTrackMinusEq("userBalance", feeAdjustedUnits, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", units, join.storedBalance());
```

Comment:
Here the fees would have been subtracted prior to calling `join.join()`, so you can skip the fee calculation and instead track the `tether.balanceOf(address(join))` as `amountReceived`. That tells you how much the `join` received (because in the harness it might already have some unaccounted balance).

Then your only assert is:
```
assertTrackPlusEq("storedBalance", amountReceived, join.storedBalance());
```